packages/material-ui/src/MenuItem/MenuItem.js

```diff
@@ -65,10 +129,47 @@ const MenuItem = React.forwardRef(function MenuItem(props, ref) {
         },
         className,
       )}
-      ref={ref}
+      onKeyDown={createChainedFunction(handleArrowRightKeydown, onKeyDown)}
+      ref={handleRef}
+      aria-expanded={subMenu ? openSubMenu : undefined}
+      aria-haspopup={subMenu ? true : undefined}
       {...other}
-    />
+    >
+      {subMenu ? (
+        <div className={classes.subMenuItemWrapper}>
+          {children}
+          <SubMenuIcon
+            className={clsx(classes.subMenuIcon, {
+              [classes.rtlSubMenuIcon]: theme.direction === 'rtl',
+            })}
+          />
+        </div>
+      ) : (
+        children
+      )}
+    </ListItem>
   );
+
+  if (!subMenu) return listItem;
```

Comment:
We're branching so often in the `listItem` element that it becomes more readable to create separate elements instead. DRY code is not desired if it's unreadable.

Reply:
I think I know what you mean, but could you elaborate on what you expect the separate elements to look like, so I'm more clear?

Reply:
I guess I'm not entirely sure if you expect separate components in separate files for each return (I don't think that's what you're saying), or if you imagine some reorganizing inside the MenuItem so the returned composition is more intuitive.

Reply:
> I guess I'm not entirely sure if you expect separate components in separate files for each return (I don't think that's what you're saying)

This is actually what I'm saying. Though I'll have to look in more detail how this would look. From what I've seen, the code tries too hard to be "DRY" when composing components would make more sense.

Reply:
Apologies, I'm realizing I missed several of your responses that were buried in collapsed parts of the conversation and I haven't been getting emails for all of these, it seems.

I'll take a look at this once I get past the latest issues I seem to have introduced during merge conflict resolution for the rebase on the emotion changes.